component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package batchedstore
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/aries-framework-go/pkg/common/log"
+	"github.com/hyperledger/aries-framework-go/pkg/newstorage"
+)
+
+const failFlush = "failure while flushing data: %w"
+
+var (
+	logger      = log.New("batchedstore")
+	errEmptyKey = errors.New("key cannot be empty")
+)
+
+// Provider is a newstorage.Provider that allows for data to be automatically batched.
+// It acts as a wrapper around another storage provider (typically, one that would benefit from batching).
+// TODO (#2484): Support batching across multiple stores for storage providers that inherently support this
+//  (like the EDV REST provider, which uses a single underlying database (vault) for all of its stores).
+type Provider struct {
+	underlyingProvider newstorage.Provider
+	openStores         map[string]*store
+	batchSizeLimit     int
+	sync.RWMutex
```

Comment:
This should probably be a private member, otherwise you're exposing the Lock/RLock functions to clients.

Reply:
Done